integration/client/restart_monitor_test.go

```diff
@@ -229,14 +230,74 @@ version = 2
 	t.Logf("%v: the task was restarted since %v", time.Now(), lastCheck)
 }
 
-// withRestartStatus is a copy of "github.com/containerd/containerd/runtime/restart".WithStatus.
-// This copy is needed because `go test` refuses circular imports.
-func withRestartStatus(status ProcessStatus) func(context.Context, *Client, *containers.Container) error {
-	return func(_ context.Context, _ *Client, c *containers.Container) error {
-		if c.Labels == nil {
-			c.Labels = make(map[string]string)
+func TestRestartMonitorWithOnFailurePolicy(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		// FIXME: The system cannot find the file specified
+		t.Skip("The system cannot find the file specified")
```

Comment:
This doesn't have to be Linux-specific.

Reply:
I'm sorry, this is an exception in the integration test of windows, but I'm not very familiar with windows. So I left FIXME.

Reply:
```patch
From c89435b9fbbd21261d38bca560546235cf7a4c27 Mon Sep 17 00:00:00 2001
From: Wei Fu <fuweid89@gmail.com>
Date: Wed, 6 Apr 2022 22:02:36 +0800
Subject: [PATCH] --wip-- [skip ci]

---
 integration/client/restart_monitor_test.go | 60 +++++++++++++++++++---
 1 file changed, 53 insertions(+), 7 deletions(-)

diff --git a/integration/client/restart_monitor_test.go b/integration/client/restart_monitor_test.go
index 571c0c66c..5571a1de7 100644
--- a/integration/client/restart_monitor_test.go
+++ b/integration/client/restart_monitor_test.go
@@ -19,6 +19,7 @@ package client
 import (
 	"bytes"
 	"context"
+	"errors"
 	"fmt"
 	"os"
 	"path/filepath"
@@ -29,11 +30,13 @@ import (
 	"time"
 
 	. "github.com/containerd/containerd"
+	eventtypes "github.com/containerd/containerd/api/events"
 	"github.com/containerd/containerd/oci"
 	"github.com/containerd/containerd/pkg/testutil"
 	"github.com/containerd/containerd/runtime/restart"
 	srvconfig "github.com/containerd/containerd/services/server/config"
 	"github.com/containerd/containerd/sys"
+	"github.com/containerd/typeurl"
 	exec "golang.org/x/sys/execabs"
 )
 
@@ -231,12 +234,8 @@ version = 2
 }
 
 func TestRestartMonitorWithOnFailurePolicy(t *testing.T) {
-	if runtime.GOOS == "windows" {
-		// FIXME: The system cannot find the file specified
-		t.Skip("The system cannot find the file specified")
-	}
 	const (
-		interval = 10 * time.Second
+		interval = 5 * time.Second
 	)
 	configTOML := fmt.Sprintf(`
 version = 2
@@ -264,7 +263,7 @@ version = 2
 		WithNewSpec(
 			oci.WithImageConfig(image),
 			// always exited with 1
-			oci.WithProcessArgs("/bin/sh", "-c", "false"),
+			withExitStatus(1),
 		),
 		restart.WithStatus(Running),
 		restart.WithPolicy(policy),
@@ -282,22 +281,69 @@ version = 2
 	if err != nil {
 		t.Fatal(err)
 	}
+
+	statusCh, err := task.Wait(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
 	defer func() {
 		if _, err := task.Delete(ctx, WithProcessKill); err != nil {
 			t.Logf("failed to delete task: %v", err)
 		}
 	}()
 
+	eventCh, eventErrCh := client.Subscribe(ctx, `topic=="/tasks/create"`)
+
 	if err := task.Start(ctx); err != nil {
 		t.Fatal(err)
 	}
-	time.Sleep(interval * 3)
+
+	select {
+	case <-statusCh:
+	case <-time.After(30 * time.Second):
+		t.Fatal("should receive exit event in time")
+	}
+
+	select {
+	case e := <-eventCh:
+		cid, err := convertTaskCreateEvent(e.Event)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if cid != id {
+			t.Fatalf("expected task id = %s, but got %s", id, cid)
+		}
+	case err := <-eventErrCh:
+		t.Fatalf("unexpected error from event channel: %v", err)
+	case <-time.After(1 * time.Minute):
+		t.Fatal("should receive task create event in time")
+	}
+
 	labels, err := container.Labels(ctx)
 	if err != nil {
 		t.Fatal(err)
 	}
+
 	restartCount, _ := strconv.Atoi(labels[restart.CountLabel])
 	if restartCount != 1 {
 		t.Fatalf("expected restart count to be 1, got %d", restartCount)
 	}
 }
+
+func convertTaskCreateEvent(e typeurl.Any) (string, error) {
+	id := ""
+
+	evt, err := typeurl.UnmarshalAny(e)
+	if err != nil {
+		return "", fmt.Errorf("failed to unmarshalany: %w", err)
+	}
+
+	switch e := evt.(type) {
+	case *eventtypes.TaskCreate:
+		id = e.ContainerID
+	default:
+		return "", errors.New("unsupported event")
+	}
+	return id, nil
+}
-- 
2.25.1

```

Try to handle the windows case and suggest to use status and sync instead of sleep(3 * time.Second) to prevent from flaky case.